docs/export-keyword.md

```diff
@@ -0,0 +1,126 @@
+# Export Keyword
+
+## Summary
+
+Allow users to export methods and values from their libraries through the use of the export keyword.
+
+## Motivation
+
+Users can export types from their libraries through the use of the export keyword.
+
+```luau
+export type Point = { x: number, y: number }
+```
+
+Consumers of these libraries are able to access exported types by indexing the libraries identifier.
+
+```luau
+local Library = require("Library")
+local point : Library.Point = { x = 0, y = 0 }
+```
+
+Right now, this is the only use of the export keyword but it would be great if we let users export more with it to offer more utility and make exporting library methods and values easier.
+
+## Design
+
+### Syntax
+You can prefix any non-local declaration with the export keyword in the top level scope of a module. For values this looks like:
+
+```luau
+export version = "5.1"
+```
+
+While methods use:
+
+```luau
+export function init()
+  -- do a thing
+end
+```
+
+### Behavior
+
+When a value or method is prefixed with export, it is automatically added to a hidden export table which is frozen as soon as the module returns.
+
+Take the following module:
+
+```luau
+export type Point = { x: number, y: number }
+
+export function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+```
+
+This essentially becomes sugar for:
+
+```luau
+local _EXT = {}
+
+type Point = { x: number, y: number }
+type _EXT.Point = Point -- note: this doesn't actually work today but one day it should
+
+local function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+_EXT.distance = distance
+
+return table.freeze(_EXT)
+```
+
+If the user attempts to assign to an exported identifier then we would throw an error explaining that the interface cannot be changed once it has been exported.
+
+### Nuances
+Due to the implementation, most things should "just-work". Here are some examples to consider:
+
+#### Calling an Exported Function
+You can call an exported function as it's registered as a local before being added to the export table.
+
+```luau
+export function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+
+distance({0, 0}, {1, 1})
+```
+
+#### Nested Tables
+You can export tables with additional values inside of them.
+
+```luau
+export triangle = {}
+
+function triangle.draw()
+
+end
+```
+
+Something important to note here is that the nested table, `triangle` is not frozen.
+
+#### Multiple Returns
+Today, you can use the export keyword along with a return statement at the end of your module. If you use the export keyword with a value however we will throw an error if you also attempt to return.
+
+```luau
+export function distance(a: Point, b: Point)
+  local x, y = a.X - b.X, a.Y - b.Y
+  return math.sqrt(x * x + y * y)
+end
+
+return table.freeze {
+  distance = distance
+}
+```
```

Comment:
Perhaps worth mentioning that the future change that would enable
`type _EXT.Point = Point -- note: this doesn't actually work today but one day it should` would also likely necessitate this restriction on explicit returning as well, since the way we'd likely be able to make that possible is by giving tables the ability to have (phantom, i.e. non-existent at runtime) types associated with them as a value.

Reply:
To me that implies we need to make this functional (even if it's bad). We can't silently drop the second return, and we can't let it override the returned table. If modules supported multiple returns we might be able to get around this by making sure the export table is always the last value returned but I suspect @vegorov-rbx might have opinions on that.

Reply:
Sorry, maybe I wasn't clear. I think the correct behavior here is that we should throw an error on any explicit return in a module with `export function` or `export id`, and I also think that even though this may seem a bit odd right now since the restriction is specific to exporting values, it's also a restriction we will have if we do the work to enable assigning a type into a "field" (not one that exists at runtime) as you had written in the earlier example.